File: src/main/scala/im/tox/toktok/app/ProfileActivity/RejectedCallAdapter.scala
package im.tox.toktok.app.ProfileActivity

import android.support.v7.widget.RecyclerView
import android.support.v7.widget.RecyclerView.ViewHolder
import android.support.v7.widget.helper.ItemTouchHelper
import android.support.v7.widget.helper.ItemTouchHelper.Callback
import android.util.Log
import android.view.View.OnTouchListener
import android.view.{MotionEvent, LayoutInflater, View, ViewGroup}
import android.widget.TextView
import im.tox.toktok.R

import scala.collection.mutable.ListBuffer

class RejectedCallAdapter(list: ListBuffer[String],dragStart: DragStart) extends RecyclerView.Adapter[RejectedCallViewHolder] with DragInterface {

  private val items: ListBuffer[String] = list

  def onCreateViewHolder(viewGroup: ViewGroup, viewType: Int): RejectedCallViewHolder = {
    val itemView: View = LayoutInflater.from(viewGroup.getContext).inflate(R.layout.rejected_call_item, viewGroup, false)
    return new RejectedCallViewHolder(itemView)
  }

  def onBindViewHolder(viewHolder: RejectedCallViewHolder, position: Int) = {
    val item: String = items(position)
    viewHolder.mMessage.setText(item)

    viewHolder.itemView.setOnTouchListener(new OnTouchListener {
      override def onTouch(v: View, event: MotionEvent): Boolean = {

        if(event.getAction == MotionEvent.ACTION_DOWN){
          dragStart.onDragStart(viewHolder)

        }

        return false
      }
    })

  }

  def getItemCount(): Int = {
    return items.length
  }

  def onItemMove(originalPosition: Int, newPosition: Int) {
    Log.d("asd","asd")
    val originalItem = items(originalPosition)
    items(originalPosition) = items(newPosition)
    items(newPosition) = originalItem
    notifyItemMoved(originalPosition, newPosition)

  }

}


final class RejectedCallViewHolder(itemView: View) extends RecyclerView.ViewHolder(itemView){

  val mMessage = itemView.findViewById(R.id.reject_item_message).asInstanceOf[TextView]

}

class DragHelperCallback(adapter: RejectedCallAdapter) extends ItemTouchHelper.Callback {

  override def isLongPressDragEnabled(): Boolean = {
    return true
  }

  override def isItemViewSwipeEnabled(): Boolean = {
    return false
  }

  override def getMovementFlags(recyclerView: RecyclerView, viewHolder: ViewHolder): Int = {

    val dragFlags = ItemTouchHelper.UP | ItemTouchHelper.DOWN;
    return Callback.makeMovementFlags(dragFlags, 0)

  }

  override def onMove(recyclerView: RecyclerView, viewHolder: ViewHolder, target: ViewHolder): Boolean = {
    Log.d("asd","asdasd")
    adapter.onItemMove(viewHolder.getAdapterPosition, target.getAdapterPosition)
    return true
  }

  def onSwiped(viewHolder: ViewHolder, direction: Int): Unit ={
  }


}

trait DragInterface{
  def onItemMove(originalPosition: Int, newPosition: Int)
}

trait DragStart{
  def onDragStart(viewHolder : ViewHolder)
}

File: src/main/java/im/tox/toktok/app/profile/RejectedCallAdapter.java
package im.tox.toktok.app.profile;

import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.RecyclerView.ViewHolder;
import android.support.v7.widget.helper.ItemTouchHelper;
import android.support.v7.widget.helper.ItemTouchHelper.Callback;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.View.OnTouchListener;
import android.view.ViewGroup;
import android.widget.RelativeLayout;
import android.widget.TextView;

import java.util.List;

import im.tox.toktok.R;

final class RejectedCallAdapter
        extends RecyclerView.Adapter<RejectedCallViewHolder>
        implements DragInterface {

    private final List<String> items;
    private final DragStart dragStart;

    RejectedCallAdapter(List<String> items, DragStart dragStart) {
        this.items = items;
        this.dragStart = dragStart;
    }

    @Override
    public RejectedCallViewHolder onCreateViewHolder(ViewGroup viewGroup, int viewType) {
        RelativeLayout itemView = (RelativeLayout) LayoutInflater.from(viewGroup.getContext()).inflate(
                R.layout.rejected_call_item, viewGroup, false);
        return new RejectedCallViewHolder(itemView);
    }

    @Override
    public void onBindViewHolder(final RejectedCallViewHolder viewHolder, int position) {
        String item = items.get(position);
        viewHolder.mMessage.setText(item);

        viewHolder.itemView.setOnTouchListener(new OnTouchListener() {
            @Override
            public boolean onTouch(View v, MotionEvent event) {
                if (event.getAction() == MotionEvent.ACTION_DOWN) {
                    dragStart.onDragStart(viewHolder);
                }

                return false;
            }
        });
    }

    @Override
    public int getItemCount() {
        return items.size();
    }

    @Override
    public void onItemMove(int originalPosition, int newPosition) {
        String originalItem = items.get(originalPosition);
        items.set(originalPosition, items.get(newPosition));
        items.set(newPosition, originalItem);
        notifyItemMoved(originalPosition, newPosition);
    }

}

final class RejectedCallViewHolder extends RecyclerView.ViewHolder {
    final TextView mMessage;

    RejectedCallViewHolder(RelativeLayout itemView) {
        super(itemView);
        mMessage = itemView.findViewById(R.id.reject_item_message);
    }
}

final class DragHelperCallback extends ItemTouchHelper.Callback {

    private final RejectedCallAdapter adapter;

    DragHelperCallback(RejectedCallAdapter adapter) {
        this.adapter = adapter;
    }

    @Override
    public boolean isLongPressDragEnabled() {
        return true;
    }

    @Override
    public boolean isItemViewSwipeEnabled() {
        return false;
    }

    @Override
    public int getMovementFlags(RecyclerView recyclerView, ViewHolder viewHolder) {
        int dragFlags = ItemTouchHelper.UP | ItemTouchHelper.DOWN;
        return Callback.makeMovementFlags(dragFlags, 0);
    }

    @Override
    public boolean onMove(RecyclerView recyclerView, ViewHolder viewHolder, ViewHolder target) {
        adapter.onItemMove(viewHolder.getAdapterPosition(), target.getAdapterPosition());
        return true;
    }

    @Override
    public void onSwiped(ViewHolder viewHolder, int direction) {
    }

}

interface DragInterface {
    void onItemMove(int originalPosition, int newPosition);
}

interface DragStart {
    void onDragStart(ViewHolder viewHolder);
}
